workers/unity/Packages/io.improbable.gdk.core/NetworkStats/NetStats.cs

```diff
@@ -0,0 +1,165 @@
+using System;
+using System.Collections.Generic;
+
+namespace Improbable.Gdk.Core.NetworkStats
+{
+    public class NetStats
+    {
+        private readonly int sequenceLength;
+        private readonly Dictionary<MessageTypeUnion, DataSequence> Data = new Dictionary<MessageTypeUnion, DataSequence>();
+        private readonly float[] frameTimes;
+
+        private int nextFrameIndex;
+
+        public NetStats(int sequenceLength)
+        {
+            this.sequenceLength = sequenceLength;
+            nextFrameIndex = sequenceLength - 1;
+
+            frameTimes = new float[sequenceLength];
+        }
+
+        public void SetFrameStats(NetFrameStats frameData, Direction direction)
+        {
+            // First we need to zero out the data before processing the _sparse_ data in the NetFrameStats.
+            foreach (var pair in Data)
+            {
+                pair.Value.Clear(nextFrameIndex, direction);
+            }
+
+            foreach (var pair in frameData.Messages)
+            {
+                if (!Data.TryGetValue(pair.Key, out var data))
+                {
+                    data = new DataSequence(sequenceLength);
+                }
+
+                data.AddFrame(nextFrameIndex, pair.Value, direction);
+                Data[pair.Key] = data;
+            }
+        }
+
+        public void SetFrameTime(float dt)
+        {
+            frameTimes[nextFrameIndex] = dt;
+        }
+
+        public void FinishFrame()
+        {
+            if (--nextFrameIndex == -1)
+            {
+                nextFrameIndex += sequenceLength;
+            }
+        }
+
+        public (DataPoint, float) GetSummary(MessageTypeUnion messageType, int numFrames, Direction direction)
+        {
+            float totalTime = 0.0f;
+
+            for (var i = 1; i <= numFrames; i++)
+            {
+                var index = (nextFrameIndex + i) % sequenceLength;
```

Comment:
what exactly is happening here

Reply:
Discussed offline - iterating through the ring buffer, will add comments to clarify in the code